sql/catalyst/src/main/scala/org/apache/spark/sql/types/StructType.scala

```diff
@@ -469,9 +469,16 @@ object StructType extends AbstractDataType {
           case leftField @ StructField(leftName, leftType, leftNullable, _) =>
             rightMapped.get(leftName)
               .map { case rightField @ StructField(_, rightType, rightNullable, _) =>
-                leftField.copy(
-                  dataType = merge(leftType, rightType),
-                  nullable = leftNullable || rightNullable)
+                Try {
+                  merge(leftType, rightType)
+                } match {
+                  case Success(dataType) =>
+                    leftField.copy(
+                      dataType = dataType,
+                      nullable = leftNullable || rightNullable)
+                  case Failure(e) =>
+                    throw new SparkException(s"Failed to merge field '$leftName': " + e.getMessage)
```

Comment:
Other exceptions in this class are also SparkException, for example the precision conflicts. Should we keep it as SparkException?
For "with both sides, `left` and `right`", do you mean just to modify the message a bit to include both left and right names(though they are the same)?
@gatorsmile your other comments are resolved.